verilog/tools/kythe/kythe_facts_extractor.cc

```diff
@@ -0,0 +1,213 @@
+#include "kythe_schema_constants.h"
+
+namespace verilog {
+namespace kythe {
+
+std::string GetFilePathFromRoot(const IndexingFactNode& root) {
+  return root.Value().Anchors()[0].Value();
+}
+
+void ExtractKytheFacts(const IndexingFactNode& node) {
+  KytheFactsExtractor kythe_extractor(GetFilePathFromRoot(node));
+  kythe_extractor.Visit(node);
+}
+
+void KytheFactsExtractor::Visit(const IndexingFactNode& node) {
+  const auto tag =
+      static_cast<IndexingFactType>(node.Value().GetIndexingFactType());
+
+  VName vname(file_path_);
+
+  switch (tag) {
+    case IndexingFactType::kFile: {
+      vname = ExtractFileFact(node);
+      break;
+    }
+    case IndexingFactType::kModule: {
+      vname = ExtractModuleFact(node);
+      break;
+    }
+    case IndexingFactType::kModuleInstance: {
+      vname = ExtractModuleInstanceFact(node);
+      break;
+    }
+    case IndexingFactType::kVariableDefinition:
+      vname = ExtractVariableDefinition(node);
+      break;
+    case IndexingFactType::kVariableReference:
+      vname = ExtractVariableReference(node);
+      break;
+  }
+
+  if (tag != IndexingFactType::kFile) {
+    std::cout << Edge(vname, kEdgeChildOf, *vnames_context_.back());
+  }
+
+  const FactTreeContextAutoPop facts_tree_auto_pop(&facts_tree_context_, &node);
+  const VNameContextAutoPop vnames_auto_pop(&vnames_context_, &vname);
+  for (const verible::VectorTree<IndexingNodeData>& child : node.Children()) {
+    Visit(child);
+  }
+}
+
+VName KytheFactsExtractor::ExtractFileFact(const IndexingFactNode& node) {
+  const VName file_vname(file_path_, "", "", "");
+
+  std::cout << Fact(file_vname, kFactNodeKind, kNodeFile);
+  std::cout << Fact(file_vname, kFactText, node.Value().Anchors()[1].Value());
+
+  return file_vname;
+}
+
+VName KytheFactsExtractor::ExtractModuleFact(const IndexingFactNode& node) {
+  const auto& anchors = node.Value().Anchors();
+  const VName module_vname(file_path_,
+                           CreateModuleSignature(anchors[0].Value()));
+
+  std::cout << Fact(module_vname, kFactNodeKind, kNodeRecord);
+  std::cout << Fact(module_vname, kFactSubkind, kSubkindModule);
+  std::cout << Fact(module_vname, kFactComplete, kCompleteDefinition);
+
+  const VName module_name_anchor = PrintAnchorVName(anchors[0], file_path_);
+
+  std::cout << Edge(module_name_anchor, kEdgeDefinesBinding, module_vname);
+
+  if (node.Value().Anchors().size() > 1) {
+    const VName module_end_label_anchor =
+        PrintAnchorVName(anchors[1], file_path_);
+    std::cout << Edge(module_end_label_anchor, kEdgeRef, module_vname);
+  }
+
+  return module_vname;
+}
+
+VName KytheFactsExtractor::ExtractModuleInstanceFact(
+    const IndexingFactNode& node) {
+  const auto& anchors = node.Value().Anchors();
+  const VName module_instance_vname(
+      file_path_,
+      CreateVariableSignature(anchors[1].Value(), *vnames_context_.back()));
+  const VName module_instance_anchor = PrintAnchorVName(anchors[1], file_path_);
```

Comment:
Assign `anchors[0]` and `[1]` to local reference variables with meaningful names.